src/cryptonote_core/service_node_list.cpp

```diff
@@ -0,0 +1,451 @@
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <functional>
+
+#include "ringct/rctSigs.h"
+#include "wallet/wallet2.h"
+#include "cryptonote_tx_utils.h"
+
+#include "service_node_list.h"
+
+#undef LOKI_DEFAULT_LOG_CATEGORY
+#define LOKI_DEFAULT_LOG_CATEGORY "service_nodes"
+
+namespace service_nodes
+{
+  service_node_list::service_node_list(cryptonote::Blockchain& blockchain)
+    : m_blockchain(blockchain)
+  {
+    blockchain.hook_block_added(*this);
+    blockchain.hook_blockchain_detached(*this);
+    blockchain.hook_init(*this);
+    blockchain.hook_validate_miner_tx(*this);
+  }
+
+  void service_node_list::init()
+  {
+    // TODO: Save this calculation, only do it if it's not here.
+    LOG_PRINT_L0("Recalculating service nodes list, scanning last 30 days");
+
+    m_service_nodes_last_reward.clear();
+
+    while (!m_rollback_events.empty())
+    {
+      delete m_rollback_events.front();
+      m_rollback_events.pop_front();
+    }
+
+    uint64_t current_height = m_blockchain.get_current_blockchain_height();
+    uint64_t start_height = (current_height >= STAKING_REQUIREMENT_LOCK_BLOCKS ? current_height - STAKING_REQUIREMENT_LOCK_BLOCKS : 0);
+    for (uint64_t height = start_height; height <= current_height; height += 1000)
+    {
+      std::list<std::pair<cryptonote::blobdata, cryptonote::block>> blocks;
+      if (!m_blockchain.get_blocks(height, 1000, blocks))
+      {
+        LOG_ERROR("Unable to initialize service nodes list");
+        return;
+      }
+      for (const auto& block_pair : blocks)
+      {
+        const cryptonote::block& block = block_pair.second;
+        std::list<cryptonote::transaction> txs;
+        std::list<crypto::hash> missed_txs;
+        if (!m_blockchain.get_transactions(block.tx_hashes, txs, missed_txs))
+        {
+          LOG_ERROR("Unable to get transactions for block " << block.hash);
+          return;
+        }
+        block_added_generic(block, txs);
+      }
+    }
+
+    m_rollback_events.push_back(new prevent_rollback(current_height));
```

Comment:
A unique_ptr would work here for C++11. `std::unique_ptr<rollback_event>(new prevent_rollback(current_height)) `which has the benefit of removing the delete calls.

Reply:
Nice, thanks